File: packages/engine/src/unit.ts
import units from './units.json' assert { type: "json" }

export enum UnitId {
    Summoner,
    BoomBrothers,
    Nomad,
    Enforcer,
    Rageclaws,
    FireDragon,
    Emberstrike,
    Scavenger,
    Batariel,
    BurningSpears,
    FireStalker,
    FireWorm,
    FireDancer,
    Firesworn,
    GiantSlayer,
    Gladiatrix,
    Juggernaut,
    MagmaFiend,
    MagmaHurler,
    MagmaSpore,
    MagmaTurret,
    Moloch,
    SkyfireDrake,
    Sunreaver,
    Vulcan,
}

export namespace UnitId {
    export function unit(id: UnitId): Unit {
        return UnitMap.get(id)!
    }
}

export class Unit {
    id: UnitId
    name: string
    cost: number
    damage: number
    health: number
    movement: number

    constructor(id: UnitId, name: string, cost: number, damage: number, health: number, movement: number) {
        this.id = id;
        this.name = name;
        this.cost = cost;
        this.damage = damage;
        this.health = health;
        this.movement = movement;
    }

    get fileName(): string {
        return this.name.toLowerCase().replace(' ', '_')
    }
}

export const Units: Unit[] = units.map((definition, index) => {
    const id: UnitId = index
    return new Unit(
      id,
      definition.name,
      definition.cost,
      definition.damage,
      definition.health,
      definition.movement,
    )
})

export const UnitMap = new Map<UnitId, Unit>()
Units.forEach(unit => UnitMap.set(unit.id, unit))

export class UnitCollection {
    private units: Unit[]

    constructor(units: Unit[]) {
        this.units = units
    }

    shuffle() {
        let currentIndex = this.units.length
        let randomIndex
        while (currentIndex != 0) {
            randomIndex = Math.floor(Math.random() * currentIndex)
            currentIndex--
            [this.units[currentIndex], this.units[randomIndex]] = [this.units[randomIndex], this.units[currentIndex]]
        }
    }

    draw(x: number): Unit[] {
        return this.units.splice(0, x)
    }

    getUnits(): Unit[] {
        let units: Unit[] = []
        for (let unit of this.units) {
            units = [...units]
        }
        return units
    }

    getUnitIds(): UnitId[] {
        let unitIds: UnitId[] = []
        for (let unit of this.units) {
            unitIds = [...unitIds, unit.id]
        }
        return unitIds
    }

    add(units: Unit[]) {
        this.units = [...this.units, ...units]
    }

    removeUnit(id: UnitId) {
        this.units = this.units.filter(unit => unit.id != id)
    }
}

File: Assets/Scripts/HexGame/Unit.cs
using System.Collections.Generic;
using System.Linq;

namespace HexGame {
  public enum UnitId {
    Summoner,
    BoomBrothers,
    Nomad,
    Enforcer,
    Rageclaws,
    FireDragon,
    Emberstrike,
    Scavenger,
    Batariel,
    BurningSpears,
    FireStalker,
    FireWorm,
    FireDancer,
    Firesworn,
    GiantSlayer,
    Gladiatrix,
    Juggernaut,
    MagmaFiend,
    MagmaHurler,
    MagmaSpore,
    MagmaTurret,
    Moloch,
    SkyfireDrake,
    Sunreaver,
    Vulcan
  }

  public class Unit {
    static Dictionary<string, UnitId> _idMap = new Dictionary<string, UnitId> {
      {"Summoner", UnitId.Summoner},
      {"Boom Brothers", UnitId.BoomBrothers},
      {"Nomad", UnitId.Nomad},
      {"Enforcer", UnitId.Enforcer},
      {"Rageclaws", UnitId.Rageclaws},
      {"Fire Dragon", UnitId.FireDragon},
      {"Emberstrike", UnitId.Emberstrike},
      {"Scavenger", UnitId.Scavenger},
      {"Batariel", UnitId.Batariel},
      {"Burning Spears", UnitId.BurningSpears},
      {"Fire Stalker", UnitId.FireStalker},
      {"Fire Worm", UnitId.FireWorm},
      {"Fire Dancer", UnitId.FireDancer},
      {"Firesworn", UnitId.Firesworn},
      {"Giant Slayer", UnitId.GiantSlayer},
      {"Gladiatrix", UnitId.Gladiatrix},
      {"Juggernaut", UnitId.Juggernaut},
      {"Magma Fiend", UnitId.MagmaFiend},
      {"Magma Hurler", UnitId.MagmaHurler},
      {"Magma Spore", UnitId.MagmaSpore},
      {"Magma Turret", UnitId.MagmaTurret},
      {"Moloch", UnitId.Moloch},
      {"Skyfire Drake", UnitId.SkyfireDrake},
      {"Sunreaver", UnitId.Sunreaver},
      {"Vulca", UnitId.Vulcan},
    };

    public string name;
    public int cost;
    public int damage;
    public int health;
    public int speed;

    public bool HasId(UnitId id) {
      return Unit.IdToString(id) == name;
    }

    public static string IdToString(UnitId id) {
      return _idMap.FirstOrDefault(x => x.Value == id).Key;
    }

    public static UnitId StringToId(string name) {
        return _idMap.TryGetValue(name, out UnitId id) ? id : UnitId.Summoner;
    }
  }
}